### ISSUE TITLE
Slow Adam sparse updates in distributed TF

### ISSUE THREAD

**Issue by Bobrosoft98** [tensorflow/tensorflow#6460]
I am trying to train a model with the **tf.nn.embedding_lookup_sparse** operation. Small example: https://gist.github.com/Bobrosoft98/2d639d3924dfbc4ec7bc620fd5a4d480

When I run this code with NUM_WORKERS = 1, the output is as follows

```
0 calc in 0.0176000595093
0 apply in 0.184364080429
0 calc in 0.0167639255524
0 apply in 0.189659118652
...
```

However, when I increase the number of workers to 30, every single process works more than 30 times slower:

```
6 calc in 0.432843923569
11 calc in 0.787642002106
3 calc in 0.440953016281
14 calc in 0.377243995667
20 calc in 0.569782018661
...
6 apply in 5.63959908485
```

The CPU load is only ~50%, so there is a lot of resources available for the computation. This makes me think that sparse updates use locking, even though the **use_locking** flag is set to False by default. There is no such problem with other optimizers (I tried GradientDescentOptimizer and AdadeltaOptimizer). Also if I exclude sparse operations from the graph (commented lines), the problem disappears.

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

https://github.com/tensorflow/tensorflow/issues/464 - there was mentioned that Adam was slower on sparse updates in general

### Environment info
Operating System: Ubuntu 16.04

Installed version of CUDA and cuDNN: I used the CPU version

If installed from binary pip package, provide:

1. A link to the pip package you installed: https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.12.0-cp27-none-linux_x86_64.whl
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
```
$ python -c "import tensorflow; print(tensorflow.__version__)"
0.12.0-rc1
```

**Comment by michaelisard** [comment]
@theweiho can you comment?

**Comment by theweiho** [comment]
This seems to be more a sparse tensor/Adam interaction issue than an embedding issue. @concretevitamin - any idea about the sparse updates part?

Or @vrv for general performance issues - or any idea who might know more about the Adam optimizer?

**Comment by concretevitamin** [comment]
+cc @mrry in case this could be solved by your recent optimizations on distributed runtime performance!

**Comment by Bobrosoft98** [comment]
I tried profiling the code as described [here](http://stackoverflow.com/a/37774470) and got interesting results.
This is what I get with GradientDescentOptimizer and 30 workers:
<img width="1010" alt="2017-01-11 21 22 17" src="https://cloud.githubusercontent.com/assets/9952727/21861086/6a3148f8-d844-11e6-89b9-a1bdef990be9.png">
This is AdamOptimizer with 30 workers
<img width="1048" alt="2017-01-11 21 22 55" src="https://cloud.githubusercontent.com/assets/9952727/21861026/30a21568-d844-11e6-8933-329dc9e7f143.png">
And this is Adam with 1 worker
<img width="908" alt="2017-01-11 21 23 03" src="https://cloud.githubusercontent.com/assets/9952727/21861027/30a45508-d844-11e6-9bd2-5a7149ca7bcb.png">
So all the expensive operations are running on the PS.

**Comment by mrry** [comment]
@concretevitamin Thanks for the hopefulness, but alas I don't think any of my recent optimizations will make this better.

@Bobrosoft98 What is your CPU utilization like on the PS machine when running Adam with one worker? From a quick glance it looks like the ops duration might be dilating because of contention for the threadpools (in particular the intra-op parallelism) that do parallel work in a TF server. If 30 workers all attempts to send their gradients at approximately the same time, there will be 30x the amount of work to do at the PS.

One possible mitigation would be to use `tf.train.SyncReplicasOptimizer` to wrap the Adam optimizer. This would have the benefit that all of the gradients from the different workers would be aggregated (relatively cheaply) before the Adam update rule is evaluated. This should cut down on the amount of work, compared to an asynchronous setup that processes the incoming batches separately. (It's not a free lunch... the synchronous replication is more vulnerable to stragglers, and a little more tricky to set up... but it might be worth it for your case.)

**Comment by Bobrosoft98** [comment]
@mrry I run everything on a single machine, the CPU load with one worker is the same as with 30 workers (about 50% on all cores).
Still, it's not clear for me why Adam has this problem, and other optimizers don't. What do mean these huge "Mul" and "Assign" blocks on the timeline? Why assigning something takes so much time?

**Comment by Taygrim** [comment]
It seems to me that a problem is adam not working properly with sparse updates. Although it is said that 
>     Note that in dense implement of this algorithm, m_t, v_t and variable will 
>     update even if g is zero, but in sparse implement, m_t, v_t and variable 
>     will not update in iterations g is zero
the code of **_apply_sparse** looks like it is updating the whole matrix.

**Comment by binxuan** [comment]
I also met the same problem when using Adam optimizer solving a model involving big embedding matrix. Compared to Adagrad or Adadelta, Adam works much slower. When I doubled the vocabulary size, the time consumption for Adam also doubled...

**Comment by binxuan** [comment]
There is a post in Tensorflow mail list that solves my problem: https://groups.google.com/a/tensorflow.org/forum/#!searchin/discuss/adam$20sparse/discuss/6GvpMz8kb-U/FaBAJkbvEQAJ

**Comment by yaroslavvb** [comment]
LazyAdamOptimizer is supposed to make it faster -- github.com/tensorflow/tensorflow/commit/819a690d

**Comment by mpjlu** [comment]
SparseAdam in pytorch is same as Tensorflow LazyAdam. The convergence of LazyAdam is lower than Adam.  The main problem is if the gradient is 0, we don't update m, and v. 
Actually, m and v should be updated when other weights are updated.  This is why LazyAdam convergence is low.
I have proposed a method to fix this issue. The performance is about the same as LazyAdam, and the convergence is the same as Adam.
Could you please review my method: 
https://github.com/tensorflow/tensorflow/issues/23668
